src/alpaka/AlpakaCore/CachingAllocator.h

```diff
@@ -0,0 +1,424 @@
+              << block.event->m_spEventImpl.get() << " .\n\t\t " << cachedBlocks_.size() << " available blocks cached ("
+              << cachedBytes_.free << " bytes), " << liveBlocks_.size() << " live blocks (" << cachedBytes_.live
+              << " bytes) outstanding." << std::endl;
+          std::cout << out.str() << std::endl;
+        }
+      }
+    }
+
+  private:
+    struct BlockDescriptor {
+      std::optional<Buffer> buffer;
+      std::optional<Queue> queue;
+      std::optional<Event> event;
+      size_t bytes = 0;
+      size_t requested = 0;  // for monitoring only
+      unsigned int bin = 0;
+
+      // the "synchronisation device" for this block
+      auto device() { return alpaka::getDev(*queue); }
+    };
+
+  private:
+    // return the maximum amount of memory that should be cached on this device
+    size_t cacheSize(size_t maxCachedBytes, double maxCachedFraction) const {
+      // note that getMemBytes() returns 0 if the platform does not support querying the device memory
+      size_t totalMemory = alpaka::getMemBytes(device_);
+      size_t memoryFraction = static_cast<size_t>(maxCachedFraction * totalMemory);
+      size_t size = std::numeric_limits<size_t>::max();
+      if (maxCachedBytes > 0 and maxCachedBytes < size) {
+        size = maxCachedBytes;
+      }
+      if (memoryFraction > 0 and memoryFraction < size) {
+        size = memoryFraction;
+      }
+      return size;
+    }
+
+    // return (bin, bin size)
+    std::tuple<unsigned int, size_t> findBin(size_t bytes) const {
+      if (bytes < minBinBytes_) {
+        return std::make_tuple(minBin_, minBinBytes_);
+      }
+      if (bytes > maxBinBytes_) {
+        throw std::runtime_error("Requested allocation size " + std::to_string(bytes) +
+                                 " bytes is too large for the caching detail with maximum bin " +
+                                 std::to_string(maxBinBytes_) +
+                                 " bytes. You might want to increase the maximum bin size");
+      }
+      unsigned int bin = minBin_;
+      size_t binBytes = minBinBytes_;
+      while (binBytes < bytes) {
+        ++bin;
+        binBytes *= binGrowth_;
+      }
+      return std::make_tuple(bin, binBytes);
+    }
+
+    bool tryReuseCachedBlock(BlockDescriptor& block) {
+      std::scoped_lock lock(mutex_);
+
+      // iterate through the range of cached blocks in the same bin
+      const auto [begin, end] = cachedBlocks_.equal_range(block.bin);
+      for (auto iBlock = begin; iBlock != end; ++iBlock) {
+        if (
+            // FIXME figure out which combination of devices allow reusing non-complete events in the same queue
+            // e.g.
+            //     for pinned host memory a freed block without completed event can not be re-used
+            //     even for operations in the same queue (?)
+            /* (*block.queue == *(iBlock->second.queue) or */
```

Comment:
@makortel could you remind me why this is disabled for the host case ?

Reply:
If enabled it could result in data race, toy example:
```cpp
auto d_buf = make_device_buffer<Foo>(queue);
{
  auto h_buf = make_host_buffer<Foo>(queue);
  // fill h_buf
  alpaka::memcpy(queue, d_buf, h_buf);
  // h_buf goes out of scope and the memory block is returned to the cache
}
{
  auto h_buf = make_host_buffer<Foo>(queue);
  // h_buf can be the same memory block as in earlier block
  // writing to h_buf before the memcpy() (or any other operation) has completed results in data race
}
```

Reply:
I see.

The same does not happen for a device buffer because all operations there are only scheduled along the queue, right ?

Reply:
> The same does not happen for a device buffer because all operations there are only scheduled along the queue, right ?

Right.